src/wasm/wasm-type.cpp

```diff
@@ -284,6 +284,63 @@ struct FiniteShapeEquator {
   bool eq(const Rtt& a, const Rtt& b);
 };
 
+// A wrapper around a RecGroup that provides equality and hashing based on the
+// structure of the group such that isorecursively equivalent recursion groups
+// will compare equal and will have the same hash. Assumes that all recursion
+// groups reachable from this one have been canonicalized, except for the
+// wrapped group itself.
+struct RecGroupStructure {
```

Comment:
Thinking on this, what do you think of the following design: Make this a class `RecGroupComparer(existing, new)` which represents a series of comparisons between two rec groups, one existing (and canonicalized) and one new (and in the process of being compared for purposes of canonicalization). This is similar to the current class except that `other` is passed in the constructor, and so it is not passed in any of the methods. I think this might be easier to understand, at least for me, and would be slightly simpler code as well. The move from "wrapping a rec group" to "comparing two rec groups" also makes things clearer since "wrapper" is less obvious I think.